src/services/refactors/convertStringOrTemplateLiteral.ts

```diff
@@ -0,0 +1,190 @@
+        }
+    }
+
+    function templateSpanToExpressions(templateSpan: TemplateSpan): Expression[] {
+        const { expression, literal } = templateSpan;
+        const text = literal.text;
+        return text.length === 0 ? [expression] : [expression, createStringLiteral(text)];
+    }
+
+    function isNotEqualsOperator(node: BinaryExpression) {
+        return node.operatorToken.kind !== SyntaxKind.EqualsToken;
+    }
+
+    function getParentBinaryExpression(expr: Node) {
+        while (isBinaryExpression(expr.parent) && isNotEqualsOperator(expr.parent)) {
+            expr = expr.parent;
+        }
+        return expr;
+    }
+
+    function arrayToTree(nodes: ReadonlyArray<Expression>, accumulator?: BinaryExpression): BinaryExpression {
+        if (nodes.length === 0) return accumulator!;
+
+        if (!accumulator) {
+            const left = nodes[0];
+            const right = nodes[1];
+
+            const binary = createBinary(left, SyntaxKind.PlusToken, right);
+            return arrayToTree(nodes.slice(2), binary);
+        }
+
+        const right = nodes[0];
+        const binary = createBinary(accumulator, SyntaxKind.PlusToken, right);
+        return arrayToTree(nodes.slice(1), binary);
+    }
+
+    function isStringConcatenationValid(node: Node): boolean {
+        const { containsString, areOperatorsValid } = treeToArray(node);
+        return containsString && areOperatorsValid;
+    }
+
+    function transformTreeToArray(node: Node): ReadonlyArray<Expression> {
+        return treeToArray(node).nodes;
+    }
+
+    function treeToArray(node: Node): { nodes: ReadonlyArray<Expression>, containsString: boolean, areOperatorsValid: boolean} {
+        if (isBinaryExpression(node)) {
+            const { nodes: leftNodes, containsString: leftHasString, areOperatorsValid: leftOperatorValid } = treeToArray(node.left);
+            const { nodes: rightNodes, containsString: rightHasString, areOperatorsValid: rightOperatorValid } = treeToArray(node.right);
+
+            if (!leftHasString && !rightHasString) {
+                return { nodes: [node], containsString: false, areOperatorsValid: true };
+            }
+
+            const currentOperatorValid = node.operatorToken.kind === SyntaxKind.PlusToken;
+            const areOperatorsValid = leftOperatorValid && currentOperatorValid && rightOperatorValid;
+
+            return { nodes: leftNodes.concat(rightNodes), containsString: true, areOperatorsValid };
+        }
+
+        return { nodes: [node as Expression], containsString: isStringLiteral(node), areOperatorsValid: true };
+    }
+
+    function concatConsecutiveString(index: number, nodes: ReadonlyArray<Expression>): [number, string] {
+        let text = "";
+
+        while (index < nodes.length && isStringLiteral(nodes[index])) {
+            text = text + decodeRawString(nodes[index].getText());

```

Comment:
you could cast `nodes[index]` to `StringLiteral` and use `nodes[index].text` instead of using `getText`. then `decodeRawString` becomes unnecessary i think.

Reply:
I first used as StringLiteral.
but last time @ajafff pointed out that Octal escapes are disallowed inside template literal

If I use `(nodes[index] as StringLiteral).text` then happens that:
```ts
// before
const foo = "Unicode \u0023 \u{0023} " + "Hex \x23 " + "Octal \43";

// after
const foo = `Unicode # # Hex # Octal 43`;
```
As you can see the octal escape is not supported with `.text` That's why I created a suboptimal function `decodeRawString`. It seems that the parser has a bug.

What do you think? 
Should I switch back to: `(nodes[index] as StringLiteral).text` ?
and provide a fix for the bug (as a separate PR)

Reply:
I switched back to `(nodes[index] as StringLiteral).text`

@RyanCavanaugh
Is there a reason why octal escape sequence is not interpreted from scanner (even when strict mode is off)
```ts
// before
const foo = "Unicode \u0023 \u{0023} " + "Hex \x23 " + "Octal \43";

// after
const foo = `Unicode # # Hex # Octal 43`;
```